cmd/sidecar/main.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/percona/percona-server-mysql-operator/pkg/mysql"
+)
+
+var log = logf.Log.WithName("sidecar")
+
+func main() {
+	opts := zap.Options{Development: true}
+	logf.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("/health/", func(w http.ResponseWriter, req *http.Request) {
+		fmt.Fprintf(w, "OK")
+	})
+	mux.HandleFunc("/backup/", backupHandler)
+	mux.HandleFunc("/logs/", logHandler)
+
+	log.Info("starting http server")
+	log.Error(http.ListenAndServe(":6033", mux), "http server failed")
+}
+
+func xtrabackupArgs() []string {
+	return []string{
+		"--backup",
+		"--stream=xbstream",
+		fmt.Sprintf("--user=%s", os.Getenv("BACKUP_USER")),
+		fmt.Sprintf("--password=%s", os.Getenv("BACKUP_PASSWORD")),
+	}
+}
+
+func backupHandler(w http.ResponseWriter, req *http.Request) {
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		http.Error(w, "HTTP server does not support streaming!", http.StatusInternalServerError)
+		return
+	}
+
+	path := strings.Split(req.URL.Path, "/")
+	if len(path) < 3 {
+		http.Error(w, "backup name must be provided in URL", http.StatusBadRequest)
+		return
+	}
+
+	backupName := path[2]
+	logFile, err := os.Create(filepath.Join(mysql.BackupLogDir, backupName+".log"))
+	if err != nil {
+		http.Error(w, "failed to create log file", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Connection", "keep-alive")
+
+	xtrabackup := exec.Command("xtrabackup", xtrabackupArgs()...)
+
+	stdout, err := xtrabackup.StdoutPipe()
+	if err != nil {
+		log.Error(err, "xtrabackup stdout pipe failed")
+		http.Error(w, "xtrabackup failed", http.StatusInternalServerError)
+		return
+	}
+
+	stderr, err := xtrabackup.StderrPipe()
+	if err != nil {
+		log.Error(err, "xtrabackup stderr pipe failed")
+		http.Error(w, "xtrabackup failed", http.StatusInternalServerError)
+		return
+	}
+
+	log.Info("Backup starting", "name", backupName)
+
+	if err := xtrabackup.Start(); err != nil {
+		log.Error(err, "failed to start xtrabackup command")
+		http.Error(w, "xtrabackup failed", http.StatusInternalServerError)
+		return
+	}
+
+	if _, err := io.Copy(w, stdout); err != nil {
+		log.Error(err, "failed to copy stdout")
+		http.Error(w, "buffer copy failed", http.StatusInternalServerError)
+		return
+	}
+
+	if _, err := io.Copy(logFile, stderr); err != nil {
+		log.Error(err, "failed to copy stderr")
+		http.Error(w, "log copy failed", http.StatusInternalServerError)
+		return
+	}
+
+	if err := xtrabackup.Wait(); err != nil {
+		log.Error(err, "failed waiting for xtrabackup to finish")
+		http.Error(w, "xtrabackup failed", http.StatusInternalServerError)
+		return
+	}
+
+	log.Info("Backup finished", "name", backupName)
```

Comment:
we need to add backup status for this message also the cluster name and name of namespace

Reply:
https://github.com/percona/percona-server-mysql-operator/pull/77/commits/674b356c5a29a79c9f36e982fdec301dfa726918